### PR TITLE
gRPC instrumentation

### PR DESCRIPTION
**Pull request by fallwith**
gRPC instrumentation

The agent will now instrument [gRPC](https://grpc.io/) activity performed by clients and servers that use the [grpc](https://rubygems.org/gems/grpc) RubyGem. Instrumentation is automatic and enabled by default, so gRPC users should not need to modify any existing application code or agent configuration to benefit from the instrumentation. The instrumentation makes use of distributed tracing for a comprehensive overview of all gRPC traffic taking place across multiple monitored applications. This allows you to observe your client and server activity using any service that adheres to the W3C standard.

The following new configuration parameters have been added for gRPC. All are optional.

| Configuration name | Default | Behavior |
| ----------- | ----------- | ----------- |
| `instrumentation.grpc_client` | auto | Set to 'disabled' to disable, set to 'chain' if there are module prepending conflicts |
| `instrumentation.grpc_server` | auto | Set to 'disabled' to disable, set to 'chain' if there are module prepending conflicts |
| `instrumentation.grpc.host_denylist` | "" |  Provide a comma delimited list of host regex patterns (ex: "private.com$,exception.*") |

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,19 @@
...
+
+    The following new configuration parameters have been added for gRPC. All are optional.
+
+    | Configuration name | Default | Behavior |
+    | ----------- | ----------- |----------- |
+    | `instrumentation.grpc_client` | auto | Set to 'disabled' to disable, set to 'chain' if there are module prepending conflicts |
+    | `instrumentation.grpc_server` | auto | Set to 'disabled' to disable, set to 'chain' if there are module prepending conflicts |
+    | `instrumentation.grpc.host_denylist` | "" |  Provide a comma delimited list of host regex patterns (ex: "private.com$,exception.*") |
```
Comment (kaylareopelle) [inline CHANGELOG.md:17]:
```suggestion
| Configuration name | Default | Behavior |
| ----------- | ----------- |----------- |
| `instrumentation.grpc_client` | auto | Set to 'disabled' to disable, set to 'chain' if there are module prepending conflicts |
| `instrumentation.grpc_server` | auto | Set to 'disabled' to disable, set to 'chain' if there are module prepending conflicts |
| `instrumentation.grpc.host_denylist` | "" |  Provide a comma delimited list of host regex patterns (ex: "private.com$,exception.*") |
```

Bringing the table to the beginning of the line to hopefully create the table in markdown!

File: infinite_tracing/test/test_helper.rb [hunk 2]
```diff
@@ -57,6 +57,8 @@ def with_serial_lock
   CLIENT_MUTEX.synchronize { yield }
 end
 
+# This will always be false unless you physically flip it to true
+# Setting this to true gives extra trace information
```
Comment (kaylareopelle) [inline infinite_tracing/test/test_helper.rb:61]: This change is unrelated to gRPC and was discovered while investigating the way we tested gRPC in infinite tracing, which uses the library.

File: lib/new_relic/agent/configuration/default_source.rb [hunk 3]
```diff
@@ -1086,6 +1086,32 @@ def self.enforce_fallback(allowed_values: nil, fallback: nil)
...
+          :allowed_from_server => false,
+          :description => 'Controls auto-instrumentation of gRPC clients at start up.  May be one of [auto|prepend|chain|disabled].'
+        },
+        :'instrumentation.grpc.host_denylist' => {
+          :default => [],
+          :public => true,
+          :type => Array,
+          :allowed_from_server => false,
+          :transform => DefaultSource.method(:convert_to_regexp_list),
+          :description => 'Define grpc connections you want the agent to ignore by specifying a list of patterns matching the hostname(s) of grpc servers to be ignored'
```
Comment (kaylareopelle) [inline lib/new_relic/agent/configuration/default_source.rb:1104]: This description should match the description in newrelic.yml

File: lib/new_relic/agent/instrumentation/grpc/client/chain.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+              &blk)
+
+              issue_request_with_tracing(:server_streamer, method, req, marshal, unmarshal,
+                deadline: deadline,
+                return_op: return_op,
+                parent: parent,
+                credentials: credentials,
+                metadata: metadata)
+
+              # TODO: gRPC - confirm that &blk is being invoked correctly
```
Comment (kaylareopelle) [inline lib/new_relic/agent/instrumentation/grpc/client/chain.rb:96]: Is this TODO complete?

File: lib/new_relic/agent/instrumentation/grpc/client/instrumentation.rb [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            return yield unless trace_with_newrelic?
+
+            segment = request_segment(method)
+            request_wrapper = NewRelic::Agent::Instrumentation::GRPC::Client::RequestWrapper.new(@host)
+            segment.add_request_headers(request_wrapper)
+            metadata.merge!(metadata, request_wrapper.instance_variable_get(:@newrelic_metadata))
+            grpc_message = nil
+            grpc_status = 0
+
+            NewRelic::Agent.disable_all_tracing do
```
Comment (kaylareopelle) [inline lib/new_relic/agent/instrumentation/grpc/client/instrumentation.rb:27]: Why is this method being called?

Comment (kaylareopelle) [reply lib/new_relic/agent/instrumentation/grpc/client/instrumentation.rb:27]: Looking in the codebase, I see a similar use case for its use around something noticing an error to avoid double tracing. Is that the purpose here too?

File: newrelic.yml [hunk 6]
```diff
@@ -406,6 +406,24 @@ common: &default_settings
...
+  # instrumentation.grpc_server: auto
+
+  # Specifies a list of hostname patterns separated by commas that will match
+  # gRPC hostnames that traffic is to be ignored by New Relic for.
+  # New Relic's gRPC client instrumentation will ignore traffic streamed to a
+  # host matching any of these patterns, and New Relic's gRPC server
+  # instrumentation will ignore traffic for a server running on a host whose
+  # hostname matches any of these patterns. By default, no traffic is ignored
+  # when gRPC instrumentation is itself enabled.
+  # For example, "private.com$,exception.*"
```
Comment (kaylareopelle) [inline newrelic.yml:424]: Could you make `newrelic.yml` and this match their descriptions?

**Review by kaylareopelle — CHANGES_REQUESTED** [review]

File: lib/new_relic/agent/configuration/default_source.rb (hunk 3 above)
Comment (fallwith) [reply lib/new_relic/agent/configuration/default_source.rb]: addressed with 3f5f9af98a932478ebcc264689da9bb278986607

File: lib/new_relic/agent/instrumentation/grpc/client/chain.rb [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+              &blk)
+
+              issue_request_with_tracing(:bidi_streamer, method, requests, marshal, unmarshal,
+                deadline: deadline,
+                return_op: return_op,
+                parent: parent,
+                credentials: credentials,
+                metadata: metadata)
+
+              # TODO: gRPC - confirm that &blk is being invoked correctly
```
[the remark this answers is not in the document]
Comment (fallwith) [reply lib/new_relic/agent/instrumentation/grpc/client/chain.rb:39]: No. I wonder if running the playground app with the instrumentation set to `chain` would do the trick.

File: lib/new_relic/agent/instrumentation/grpc/client/chain.rb (hunk 4 above)
Comment (fallwith) [reply lib/new_relic/agent/instrumentation/grpc/client/chain.rb:96]: No. I wonder if running the playground app with the instrumentation set to chain would do the trick.

File: lib/new_relic/agent/instrumentation/grpc/client/instrumentation.rb (hunk 5 above)
Comment (fallwith) [reply lib/new_relic/agent/instrumentation/grpc/client/instrumentation.rb:27]: This pattern was copied from `net_http/instrumentation.rb`, which was used as a reference instrumentation implementation. My hunch was that the agent ought to be completely out of the way of the original instrumented method and not inadvertently interfere with it. But the prevention of duplication makes sense.

File: newrelic.yml (hunk 6 above)
Comment (fallwith) [reply newrelic.yml:424]: `default_source.rb` addressed with 3f5f9af98a932478ebcc264689da9bb278986607

**Review by kaylareopelle — APPROVED** [review]
🍵
